ocaml/xapi/certificates.ml

```diff
@@ -334,3 +335,57 @@ let validate_private_key pkcs8_private_key =
   |> function
     | Ok priv -> priv
     | Error `Msg (err, msg) -> raise_server_error msg err
+
+let validate_certificate kind str now private_key =
+  let open Rresult in
+  let ensure_keys_match private_key certificate =
+    let public_key = X509.Certificate.public_key certificate in
+    match public_key, private_key with
+    | `RSA pub, `RSA priv when pub = Nocrypto.Rsa.pub_of_priv priv ->
```

Comment:
Is polymorphic equality the only way to compare these, or would the library provide a comparison function?

Reply:
The certificate module provides
```
val supports_keytype : t -> key_type -> bool

    supports_keytype certificate key_type is result, whether public key of the certificate matches the given key_type.
```
But I don't see a function to extract keytype from a private or public key. I don't think I can avoid polymorphic equality.